File: sdks/python/crowdcontrol/validate.py
"""Static schema validation for CrowdControl policies.

Ports github.com/mikemackintosh/crowdcontrol/evaluator/validate.go to Python.
Produces non-fatal warnings for typos, type mismatches, and other issues.
"""

from __future__ import annotations

import re

from .types import (
    FIELD_ANY,
    FIELD_BOOL,
    FIELD_LIST,
    FIELD_MAP,
    FIELD_NUMBER,
    FIELD_STRING,
    Condition,
    ConditionType,
    Expr,
    ExprKind,
    Policy,
    Schema,
    SchemaWarning,
)


def validate_policies(policies: list[Policy], schema: Schema) -> list[SchemaWarning]:
    """Run schema validation across every loaded policy."""
    warnings: list[SchemaWarning] = []
    for policy in policies:
        for rule in policy.rules:
            for cond in rule.conditions:
                warnings.extend(_validate_condition(cond, schema, rule.name))
            for u in rule.unlesses:
                warnings.extend(_validate_condition(u, schema, rule.name))
            if rule.message:
                warnings.extend(_validate_interpolations(rule.message, schema, rule.name))
    return warnings


def _validate_condition(cond: Condition, schema: Schema, rule_name: str) -> list[SchemaWarning]:
    warnings: list[SchemaWarning] = []
    t = cond.type

    if t == ConditionType.FIELD:
        if cond.field:
            warnings.extend(_check_field(cond.field, schema, rule_name, cond))
    elif t == ConditionType.HAS:
        if cond.field:
            warnings.extend(_check_field_exists(cond.field, schema, rule_name))
    elif t == ConditionType.AGGREGATE:
        if cond.aggregate_target:
            warnings.extend(_check_aggregate_field(cond.aggregate_target, schema, rule_name))
    elif t in (ConditionType.ANY, ConditionType.ALL):
        if cond.list_field:
            warnings.extend(_check_list_field(cond.list_field, schema, rule_name))
        if cond.predicate is not None:
            warnings.extend(_validate_condition(cond.predicate, schema, rule_name))
    elif t == ConditionType.OR:
        for sub in cond.or_group:
            warnings.extend(_validate_condition(sub, schema, rule_name))
    elif t == ConditionType.EXPR:
        if cond.left_expr is not None:
            warnings.extend(_check_expr_fields(cond.left_expr, schema, rule_name))
        if cond.right_expr is not None:
            warnings.extend(_check_expr_fields(cond.right_expr, schema, rule_name))

    return warnings


def _check_field(field: str, schema: Schema, rule_name: str, cond: Condition) -> list[SchemaWarning]:
    warnings: list[SchemaWarning] = []
    expected = _lookup_field(schema, field)
    if expected is None:
        warnings.append(
            SchemaWarning(
                rule=rule_name,
                field=field,
                message=f"field {field!r} not found in schema",
            )
        )
        return warnings

    op = cond.op
    if op in ("<", ">", "<=", ">="):
        if expected not in (FIELD_NUMBER, FIELD_ANY):
            warnings.append(
                SchemaWarning(
                    rule=rule_name,
                    field=field,
                    message=f"operator {op} used on field {field!r} of type {expected}",
                )
            )
    elif op in ("contains", "intersects", "is_subset"):
        if expected not in (FIELD_LIST, FIELD_STRING, FIELD_ANY):
            warnings.append(
                SchemaWarning(
                    rule=rule_name,
                    field=field,
                    message=f"operator {op} used on field {field!r} of type {expected}",
                )
            )
    elif op == "in":
        if expected not in (FIELD_STRING, FIELD_ANY):
            warnings.append(
                SchemaWarning(
                    rule=rule_name,
                    field=field,
                    message=f"operator 'in' used on field {field!r} of type {expected}",
                )
            )
    return warnings


def _check_field_exists(field: str, schema: Schema, rule_name: str) -> list[SchemaWarning]:
    if _lookup_field(schema, field) is None:
        return [
            SchemaWarning(
                rule=rule_name,
                field=field,
                message=f"field {field!r} not found in schema (used with 'has')",
            )
        ]
    return []


def _check_aggregate_field(field: str, schema: Schema, rule_name: str) -> list[SchemaWarning]:
    expected = _lookup_field(schema, field)
    if expected is None:
        return [
            SchemaWarning(
                rule=rule_name,
                field=field,
                message=f"field {field!r} not found in schema (used with 'count')",
            )
        ]
    if expected not in (FIELD_LIST, FIELD_NUMBER, FIELD_ANY):
        return [
            SchemaWarning(
                rule=rule_name,
                field=field,
                message=f"count() used on field {field!r} of type {expected}, expected list or number",
            )
        ]
    return []


def _check_list_field(field: str, schema: Schema, rule_name: str) -> list[SchemaWarning]:
    expected = _lookup_field(schema, field)
    if expected is None:
        return [
            SchemaWarning(
                rule=rule_name,
                field=field,
                message=f"field {field!r} not found in schema (used with quantifier)",
            )
        ]
    if expected not in (FIELD_LIST, FIELD_ANY):
        return [
            SchemaWarning(
                rule=rule_name,
                field=field,
                message=f"quantifier used on field {field!r} of type {expected}, expected list",
            )
        ]
    return []


def _check_expr_fields(expr: Expr, schema: Schema, rule_name: str) -> list[SchemaWarning]:
    warnings: list[SchemaWarning] = []
    k = expr.kind
    if k == ExprKind.FIELD and expr.field:
        expected = _lookup_field(schema, expr.field)
        if expected is None:
            warnings.append(
                SchemaWarning(
                    rule=rule_name,
                    field=expr.field,
                    message=f"field {expr.field!r} not found in schema (used in arithmetic)",
                )
            )
        elif expected not in (FIELD_NUMBER, FIELD_ANY):
            warnings.append(
                SchemaWarning(
                    rule=rule_name,
                    field=expr.field,
                    message=f"arithmetic used on field {expr.field!r} of type {expected}, expected number",
                )
            )
    elif k == ExprKind.COUNT and expr.agg_target:
        warnings.extend(_check_aggregate_field(expr.agg_target, schema, rule_name))
    elif k == ExprKind.LEN and expr.field:
        expected = _lookup_field(schema, expr.field)
        if expected is None:
            warnings.append(
                SchemaWarning(
                    rule=rule_name,
                    field=expr.field,
                    message=f"field {expr.field!r} not found in schema (used with len)",
                )
            )
    elif k == ExprKind.BINARY:
        if expr.left is not None:
            warnings.extend(_check_expr_fields(expr.left, schema, rule_name))
        if expr.right is not None:
            warnings.extend(_check_expr_fields(expr.right, schema, rule_name))
    return warnings


_interp_re = re.compile(r"\{([^}]+)\}")


def _validate_interpolations(msg: str, schema: Schema, rule_name: str) -> list[SchemaWarning]:
    warnings: list[SchemaWarning] = []
    for match in _interp_re.finditer(msg):
        expr = match.group(1)
        if expr.startswith("count(") and expr.endswith(")"):
            continue
        if _lookup_field(schema, expr) is None:
            warnings.append(
                SchemaWarning(
                    rule=rule_name,
                    field=expr,
                    message=f"message interpolation references unknown field {expr!r}",
                )
            )
    return warnings


def _lookup_field(schema: Schema, field: str):
    if field in schema.fields:
        return schema.fields[field]
    parts = field.split(".")
    for i in range(len(parts) - 1, 0, -1):
        prefix = ".".join(parts[:i])
        if prefix in schema.fields and schema.fields[prefix] == FIELD_MAP:
            return FIELD_ANY
    return None


def format_warnings(warnings: list[SchemaWarning]) -> str:
    if not warnings:
        return ""
    lines = [f"  {w.rule}: {w.message}" for w in warnings]
    return "\n".join(lines) + "\n"

File: sdks/ruby/lib/crowdcontrol/validate.rb
# frozen_string_literal: true

require_relative "types"

module CrowdControl
  class << self
    # Run schema validation across every loaded policy. Returns a list of
    # non-fatal SchemaWarning values.
    def validate_policies(policies, schema)
      warnings = []
      policies.each do |policy|
        policy.rules.each do |rule|
          rule.conditions.each do |cond|
            warnings.concat(validate_condition(cond, schema, rule.name))
          end
          rule.unlesses.each do |u|
            warnings.concat(validate_condition(u, schema, rule.name))
          end
          warnings.concat(validate_interpolations(rule.message, schema, rule.name)) if rule.message && rule.message != ""
        end
      end
      warnings
    end

    def format_warnings(warnings)
      return "" if warnings.nil? || warnings.empty?

      warnings.map { |w| "  #{w.rule}: #{w.message}" }.join("\n") + "\n"
    end

    # ------------------------------------------------------------------
    # Internal
    # ------------------------------------------------------------------

    def validate_condition(cond, schema, rule_name)
      warnings = []
      case cond.type
      when ConditionType::FIELD
        warnings.concat(check_field(cond.field, schema, rule_name, cond)) unless cond.field.nil? || cond.field == ""
      when ConditionType::HAS
        warnings.concat(check_field_exists(cond.field, schema, rule_name)) unless cond.field.nil? || cond.field == ""
      when ConditionType::AGGREGATE
        warnings.concat(check_aggregate_field(cond.aggregate_target, schema, rule_name)) unless cond.aggregate_target.nil? || cond.aggregate_target == ""
      when ConditionType::ANY, ConditionType::ALL
        warnings.concat(check_list_field(cond.list_field, schema, rule_name)) unless cond.list_field.nil? || cond.list_field == ""
        warnings.concat(validate_condition(cond.predicate, schema, rule_name)) if cond.predicate
      when ConditionType::OR
        cond.or_group.each { |sub| warnings.concat(validate_condition(sub, schema, rule_name)) }
      when ConditionType::EXPR
        warnings.concat(check_expr_fields(cond.left_expr, schema, rule_name))  if cond.left_expr
        warnings.concat(check_expr_fields(cond.right_expr, schema, rule_name)) if cond.right_expr
      end
      warnings
    end

    def check_field(field, schema, rule_name, cond)
      warnings = []
      expected = lookup_field(schema, field)
      if expected.nil?
        warnings << SchemaWarning.new(
          rule: rule_name,
          field: field,
          message: "field #{field.inspect} not found in schema"
        )
        return warnings
      end

      op = cond.op
      case op
      when "<", ">", "<=", ">="
        unless [FIELD_NUMBER, FIELD_ANY].include?(expected)
          warnings << SchemaWarning.new(
            rule: rule_name,
            field: field,
            message: "operator #{op} used on field #{field.inspect} of type #{expected}"
          )
        end
      when "contains", "intersects", "is_subset"
        unless [FIELD_LIST, FIELD_STRING, FIELD_ANY].include?(expected)
          warnings << SchemaWarning.new(
            rule: rule_name,
            field: field,
            message: "operator #{op} used on field #{field.inspect} of type #{expected}"
          )
        end
      when "in"
        unless [FIELD_STRING, FIELD_ANY].include?(expected)
          warnings << SchemaWarning.new(
            rule: rule_name,
            field: field,
            message: "operator 'in' used on field #{field.inspect} of type #{expected}"
          )
        end
      end
      warnings
    end

    def check_field_exists(field, schema, rule_name)
      return [] unless lookup_field(schema, field).nil?

      [SchemaWarning.new(
        rule: rule_name,
        field: field,
        message: "field #{field.inspect} not found in schema (used with 'has')"
      )]
    end

    def check_aggregate_field(field, schema, rule_name)
      expected = lookup_field(schema, field)
      if expected.nil?
        return [SchemaWarning.new(
          rule: rule_name,
          field: field,
          message: "field #{field.inspect} not found in schema (used with 'count')"
        )]
      end
      return [] if [FIELD_LIST, FIELD_NUMBER, FIELD_ANY].include?(expected)

      [SchemaWarning.new(
        rule: rule_name,
        field: field,
        message: "count() used on field #{field.inspect} of type #{expected}, expected list or number"
      )]
    end

    def check_list_field(field, schema, rule_name)
      expected = lookup_field(schema, field)
      if expected.nil?
        return [SchemaWarning.new(
          rule: rule_name,
          field: field,
          message: "field #{field.inspect} not found in schema (used with quantifier)"
        )]
      end
      return [] if [FIELD_LIST, FIELD_ANY].include?(expected)

      [SchemaWarning.new(
        rule: rule_name,
        field: field,
        message: "quantifier used on field #{field.inspect} of type #{expected}, expected list"
      )]
    end

    def check_expr_fields(expr, schema, rule_name)
      warnings = []
      case expr.kind
      when ExprKind::FIELD
        return warnings if expr.field.nil? || expr.field == ""

        expected = lookup_field(schema, expr.field)
        if expected.nil?
          warnings << SchemaWarning.new(
            rule: rule_name,
            field: expr.field,
            message: "field #{expr.field.inspect} not found in schema (used in arithmetic)"
          )
        elsif ![FIELD_NUMBER, FIELD_ANY].include?(expected)
          warnings << SchemaWarning.new(
            rule: rule_name,
            field: expr.field,
            message: "arithmetic used on field #{expr.field.inspect} of type #{expected}, expected number"
          )
        end
      when ExprKind::COUNT
        warnings.concat(check_aggregate_field(expr.agg_target, schema, rule_name)) unless expr.agg_target.nil? || expr.agg_target == ""
      when ExprKind::LEN
        return warnings if expr.field.nil? || expr.field == ""

        expected = lookup_field(schema, expr.field)
        if expected.nil?
          warnings << SchemaWarning.new(
            rule: rule_name,
            field: expr.field,
            message: "field #{expr.field.inspect} not found in schema (used with len)"
          )
        end
      when ExprKind::BINARY
        warnings.concat(check_expr_fields(expr.left, schema, rule_name))  if expr.left
        warnings.concat(check_expr_fields(expr.right, schema, rule_name)) if expr.right
      end
      warnings
    end

    def validate_interpolations(msg, schema, rule_name)
      warnings = []
      msg.scan(/\{([^}]+)\}/) do |m|
        expr = m[0]
        next if expr.start_with?("count(") && expr.end_with?(")")
        next unless lookup_field(schema, expr).nil?

        warnings << SchemaWarning.new(
          rule: rule_name,
          field: expr,
          message: "message interpolation references unknown field #{expr.inspect}"
        )
      end
      warnings
    end

    def lookup_field(schema, field)
      return schema.fields[field] if schema.fields.key?(field)

      parts = field.split(".")
      (parts.length - 1).downto(1) do |i|
        prefix = parts[0...i].join(".")
        if schema.fields.key?(prefix) && schema.fields[prefix] == FIELD_MAP
          return FIELD_ANY
        end
      end
      nil
    end
  end
end
